noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays.nr

```diff
@@ -231,121 +269,315 @@ where
+
+    fn validate_empty_rhs(array: [T; N], length_hint: u32) {
+        for_i_in_(
+            length_hint,
+            N,
+            0,
+            N,
+            for_loop_no_op,
+            |i| { array[i].assert_empty("Array should be all zeros after its 'length'"); },
+            for_loop_no_op,
+            true,
+            false,
+        );
+    }
+}
+
+impl<T, let N: u32> ClaimedLengthArray<T, N>
+where
+    T: Empty,
+{
+    // No constructor. Append to an empty one.
+
+    // For constrained append functions, see the dedicated file: assert_array_appended.nr
+
+    pub fn push(&mut self, item: T) {
+        assert(self.length != N, "Array full");
+
+        let mut next_index = self.length;
+        self.array[next_index] = item;
+        self.length += 1;
+    }
+
+    pub fn assert_dense_trimmed(self) {
+        for_i_in_0_(
+            self.length,
+            self.array.len(),
+            |i| {
+                self.array[i].assert_not_empty(); // LHS of input array is not dense.
+            },
+            |i| self.array[i].assert_empty("RHS of input array is not empty"),
+            false,
+        );
+    }
+
+    pub fn assert_empty<let S: u32>(self, msg: str<S>) {
+        for i in 0..N {
+            self.array[i].assert_empty(msg);
         }
+        assert_eq(self.length, 0);
+    }
+
+    pub unconstrained fn push_hint(&mut self, item: T) {
+        assert(self.length != N, "Array full");
+
+        let mut next_index = self.length;
+        self.array[next_index] = item;
+        self.length += 1;
+    }
+
+    pub unconstrained fn pop_hint(&mut self) -> T {
+        assert(self.length != 0, "Array empty");
+
+        let mut top_index = self.length - 1;
+        let popped_item = self.array[top_index];
+        self.array[top_index] = T::empty();
+        self.length -= 1;
+        popped_item
+    }
+
+    // E.g.
+    // dest.for_each_i(|source_item, i| { assert_eq(dest.array[i], source_item, "bad copy"); })
+    pub fn for_each_i<Env>(self, f: fn[Env](T, u32) -> ()) {
+        // We pass `false`, because by construction, we should know that self.length <= self.array.len().
+        for_i_only_in_0_(
+            self.length,
+            self.array.len(),
+            |i| f(self.array[i], i),
+            false,
+        );
+    }
+
+    pub fn from_bounded_vec(vec: BoundedVec<T, N>) -> Self {
+        Self { array: vec.storage(), length: vec.len() }
     }
-    index
 }
 
-// Routine which returns the length of an array right padded by empty elements
-// of the form: [*,*,*...,0,0,0,0] where * is any value (zeroes allowed).
-// See smoke_validate_array_trailing for examples.
-// Nothing to do with validated arrays. Correctness constrained by padded_array_length.
-pub unconstrained fn unsafe_padded_array_length<T, let N: u32>(array: [T; N]) -> u32
+// TODO: compiler bug. No idea why this is needed, if we have #[derive(Eq)] above the struct definition.
+impl<T, let N: u32> Eq for ClaimedLengthArray<T, N>
```

Comment:
If you're getting an error in metaprogramming where this can't be found then it's possible you're falling afoul of macro ordering rules.